store/storage/sqllite/db.go

```diff
@@ -0,0 +1,221 @@
+package sqllite
+
+import (
+	"bytes"
+	"database/sql"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	_ "modernc.org/sqlite"
+
+	"cosmossdk.io/store/v2"
+)
+
+const (
+	driverName       = "sqlite"
+	dbName           = "ss.db"
+	reservedStoreKey = "_RESERVED_"
+	keyLatestHeight  = "latest_height"
+
+	latestVersionStmt = `
+	INSERT INTO state_storage(store_key, key, value, version)
+    VALUES(?, ?, ?, ?)
+  ON CONFLICT(store_key, key, version) DO UPDATE SET
+    value = ?;
+	`
+	upsertStmt = `
+	INSERT INTO state_storage(store_key, key, value, version)
+    VALUES(?, ?, ?, ?)
+  ON CONFLICT(store_key, key, version) DO UPDATE SET
+    value = ?;
+	`
+	delStmt = "DELETE FROM state_storage WHERE store_key = ? AND key = ? AND version = ?;"
+)
+
+var _ store.VersionedDatabase = (*Database)(nil)
+
+type Database struct {
+	storage *sql.DB
+}
+
+func New(dataDir string) (*Database, error) {
+	db, err := sql.Open(driverName, filepath.Join(dataDir, dbName))
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sqlite DB: %w", err)
+	}
+
+	stmt := `
+	CREATE TABLE IF NOT EXISTS state_storage (
+		id integer not null primary key, 
+		store_key varchar not null,
```

Comment:
Since tree depth directly correlates with read speed, the best performance will probably come by partitioning tables on `store_key`, read one table per `store_key`.

I think we should run some realistic bench marks against RocksDB and PebbleDB first to get a baseline.  My intuition is that sqlite (or any BTree) will outperform either *given the proper configuration*, but that will be a journey of tweaking.

Reply:
Noted. But that will come at the cost of complexity of managing multiple tables. I.e. we'll need to create tables on the fly when they do not exist. I will write benchmarks in a future PR for all SS backends.